src/cmap/auth/mongodb_oidc/callback_workflow.ts

```diff
@@ -0,0 +1,209 @@
+    if (refresh) {
+      const result: OIDCRequestTokenResult = await refresh(
+        credentials.username,
+        stepOneResult,
+        tokenResult,
+        TIMEOUT
+      );
+      // Cache a new entry and continue with the saslContinue.
+      this.cache.addEntry(result, stepOneResult, connection.address, credentials.username);
+      return finishAuth(result, conversationId, connection, credentials);
+    } else {
+      // Fallback to using the request callback.
+      return this.requestAndFinish(connection, credentials, stepOneResult, conversationId);
+    }
+  }
+
+  /**
+   * Execute the request callback and finish authentication.
+   */
+  private async requestAndFinish(
+    connection: Connection,
+    credentials: MongoCredentials,
+    stepOneResult: OIDCMechanismServerStep1,
+    conversationId?: number
+  ): Promise<Document> {
+    // Call the request callback.
+    const request = credentials.mechanismProperties.REQUEST_TOKEN_CALLBACK;
+    // Always clear expired entries from the cache on each finish as cleanup.
+    this.cache.deleteExpiredEntries();
+    if (request) {
+      const tokenResult = await request(credentials.username, stepOneResult, TIMEOUT);
+      // Cache a new entry and continue with the saslContinue.
+      this.cache.addEntry(tokenResult, stepOneResult, connection.address, credentials.username);
+      return finishAuth(tokenResult, conversationId, connection, credentials);
+    } else {
+      // Request callback must be present.
+      throw new MongoInvalidArgumentError(
+        'Auth mechanism property REQUEST_TOKEN_CALLBACK is required.'
+      );
+    }
+  }
+}
+
+/**
+ * Cache the result of the user supplied callback and execute the
+ * step two saslContinue.
+ */
+async function finishAuth(
+  result: OIDCRequestTokenResult,
+  conversationId: number | undefined,
+  connection: Connection,
+  credentials: MongoCredentials
+): Promise<Document> {
+  // Execute the step two saslContinue.
+  const executeCommand = promisify(connection.command.bind(connection));
+  return executeCommand(
+    ns(credentials.source),
+    continueCommandDocument(result.accessToken, conversationId),
+    undefined
+  );
+}
+
+/**
+ * Generate the saslStart command document.
+ */
+function startCommandDocument(credentials: MongoCredentials): Document {
+  const payload: Document = {};
+  if (credentials.username) {
+    payload.n = credentials.username;
+  }
+  return {
+    saslStart: 1,
+    autoAuthorize: 1,
+    mechanism: AuthMechanism.MONGODB_OIDC,
+    payload: new Binary(BSON.serialize(payload))
+  };
+}
+
+/**
+ * Generate the saslContinue command document.
+ */
+function continueCommandDocument(token: string, conversationId?: number): Document {
+  if (conversationId) {
+    return {
+      saslContinue: 1,
+      conversationId: conversationId,
+      payload: new Binary(BSON.serialize({ jwt: token }))
+    };
+  }
+  return {
+    saslStart: 1,
```

Comment:
can we add a comment explaining why we're returning a saslStart command from `continueCommandDocument`?  I remember discussing this last thursday during tech chat but it would be nice to provide context for future devs (including myself)